File: packages/test-results/__tests__/test-results.test.ts
import { describe, it, expect } from 'vitest'
import {
  summarizeTests,
  createTestResults,
  type TestResultData,
} from '../src/index.js'

const pass = (id: string): TestResultData => ({
  id,
  name: `Test ${id}`,
  status: 'pass',
})
const fail = (id: string): TestResultData => ({
  id,
  name: `Test ${id}`,
  status: 'fail',
  expected: 'true',
  actual: 'false',
  message: 'Expected true but received false',
})
const skip = (id: string): TestResultData => ({
  id,
  name: `Test ${id}`,
  status: 'skip',
  message: 'Skipped: not implemented',
})

describe('summarizeTests', () => {
  it('counts passing tests correctly', () => {
    const summary = summarizeTests([pass('1'), pass('2'), pass('3')])
    expect(summary.total).toBe(3)
    expect(summary.passed).toBe(3)
    expect(summary.failed).toBe(0)
    expect(summary.skipped).toBe(0)
    expect(summary.allPassed).toBe(true)
  })

  it('counts failing tests and sets allPassed false', () => {
    const summary = summarizeTests([pass('1'), fail('2'), pass('3')])
    expect(summary.total).toBe(3)
    expect(summary.passed).toBe(2)
    expect(summary.failed).toBe(1)
    expect(summary.skipped).toBe(0)
    expect(summary.allPassed).toBe(false)
  })

  it('counts skipped tests and sets allPassed false', () => {
    const summary = summarizeTests([pass('1'), pass('2'), skip('3')])
    expect(summary.total).toBe(3)
    expect(summary.passed).toBe(2)
    expect(summary.failed).toBe(0)
    expect(summary.skipped).toBe(1)
    expect(summary.allPassed).toBe(false)
  })

  it('counts mixed results correctly', () => {
    const summary = summarizeTests([pass('1'), fail('2'), skip('3'), pass('4')])
    expect(summary.total).toBe(4)
    expect(summary.passed).toBe(2)
    expect(summary.failed).toBe(1)
    expect(summary.skipped).toBe(1)
    expect(summary.allPassed).toBe(false)
  })

  it('returns zero totals for an empty list', () => {
    const summary = summarizeTests([])
    expect(summary.total).toBe(0)
    expect(summary.passed).toBe(0)
    expect(summary.failed).toBe(0)
    expect(summary.skipped).toBe(0)
    expect(summary.allPassed).toBe(false)
  })
})

describe('createTestResults', () => {
  it('returns role="list" in ariaProps', () => {
    const { ariaProps } = createTestResults()
    expect(ariaProps.role).toBe('list')
  })

  it('returns data-component attribute', () => {
    const { dataAttributes } = createTestResults()
    expect(dataAttributes['data-component']).toBe('test-results')
  })
})

File: packages/flutter/test/test_results_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:refraction_ui/refraction_ui.dart';
import 'package:refraction_ui/src/components/test_results.dart';

void main() {
  Widget buildTestApp(Widget child) {
    return MaterialApp(
      home: RefractionTheme(
        data: RefractionThemeData.minimalLight(),
        child: Scaffold(body: child),
      ),
    );
  }

  testWidgets('RefractionTestResults renders summary and rows correctly', (
    WidgetTester tester,
  ) async {
    final results = [
      const TestResultData(
        id: '1',
        name: 'Test Pass',
        status: TestStatus.pass,
        durationMs: 5,
      ),
      const TestResultData(
        id: '2',
        name: 'Test Fail',
        status: TestStatus.fail,
        durationMs: 15,
        expected: 'Expected Output',
        actual: 'Actual Output',
        message: 'Stacktrace info',
      ),
      const TestResultData(
        id: '3',
        name: 'Test Skip',
        status: TestStatus.skip,
        message: 'Skipped for now',
      ),
    ];

    await tester.pumpWidget(
      buildTestApp(
        RefractionTestResults(results: results),
      ),
    );

    // Summary text
    expect(find.text('1/3 passed · 1 failed · 1 skipped'), findsOneWidget);

    // Rows
    expect(find.text('Test Pass'), findsOneWidget);
    expect(find.text('Test Fail'), findsOneWidget);
    expect(find.text('Test Skip'), findsOneWidget);

    // Badges
    expect(find.text('PASS'), findsOneWidget);
    expect(find.text('FAIL'), findsOneWidget);
    expect(find.text('SKIP'), findsOneWidget);

    // Error details inside RichText
    expect(
      find.byWidgetPredicate((widget) =>
          widget is RichText &&
          widget.text.toPlainText().contains('Expected Output')),
      findsOneWidget,
    );
    expect(
      find.byWidgetPredicate((widget) =>
          widget is RichText &&
          widget.text.toPlainText().contains('Actual Output')),
      findsOneWidget,
    );

    expect(find.text('Stacktrace info'), findsOneWidget);
    expect(find.text('Skipped for now'), findsOneWidget);
  });
}
